# Segmenting and Clustering PART 1

### Download all the dependencies that we will need.

In [3]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import json, lxml
from geopy.geocoders import Nominatim # convert an address into latitude and longitude values
import requests # library to handle requests
from pandas.io.json import json_normalize # tranform JSON file into a pandas dataframe
# Matplotlib and associated plotting modules
import matplotlib.cm as cm
import matplotlib.colors as colors
# import k-means from clustering stage
from sklearn.cluster import KMeans
# import folium # map rendering library
from bs4 import BeautifulSoup
import warnings
warnings.filterwarnings('ignore')
try:
    import folium
except:
    !pip install folium
    import folium

print('Libraries imported.')

Libraries imported.


### Reading the Data

In [5]:
url = 'https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M'
source = requests.get(url).text
soup = BeautifulSoup(source)

table_data = soup.find('div', class_='mw-parser-output')
table = table_data.table.tbody

columns = ['PostalCode', 'Borough', 'Neighbourhood']
data = dict({key:[]*len(columns) for key in columns})

for row in table.find_all('tr'):
    for i,column in zip(row.find_all('td'),columns):
        i = i.text
        i = i.replace('\n', '')
        data[column].append(i)

df = pd.DataFrame.from_dict(data=data)[columns]
print(df.shape)
df.head()

(180, 3)


PostalCode           Borough               Neighbourhood
0        M1A      Not assigned                            
1        M2A      Not assigned                            
2        M3A        North York                   Parkwoods
3        M4A        North York            Victoria Village
4        M5A  Downtown Toronto  Regent Park / Harbourfront

### Data contains some rows where value are not assigned.
### Dropping the row where Borough is Not assigned (only)

In [6]:
df = df[df['Borough'] != 'Not assigned'].reset_index(drop = True)
print('After dropping rows where borough is "Not assigned", Shape is: ',df.shape)
print('Number of rows where Neighbourhood is "Not assigned" but borough has value: ', 
      df[df['Neighbourhood'] == 'Not assigned'].shape[0])

After dropping rows where borough is "Not assigned", Shape is:  (103, 3)
Number of rows where Neighbourhood is "Not assigned" but borough has value:  0


#### There is no row where Neighbourhood is Not assigned but Borough is assigned.
#### Where Borough is not 'Not assigned' (means there a borugh) and Neighbourhood is 'Not assigned', borough will be the neighbourhood.

In [7]:
p, b, n = [], [], []
for postcode, borough, neigh in zip(df['PostalCode'], df['Borough'], df['Neighbourhood']):
    p.append(postcode)
    b.append(borough)
    if neigh == 'Not assigned':
        n.append(borough)
    else:
        n.append(neigh)

df = pd.DataFrame({'PostalCode': p, 'Borough': b, 'Neighbourhood':n})[columns]
print(df.shape)
df.head()

(103, 3)


PostalCode           Borough                                 Neighbourhood
0        M3A        North York                                     Parkwoods
1        M4A        North York                              Victoria Village
2        M5A  Downtown Toronto                    Regent Park / Harbourfront
3        M6A        North York             Lawrence Manor / Lawrence Heights
4        M7A  Downtown Toronto  Queen's Park / Ontario Provincial Government

#### 77 rows has been dropped because there was no assigned value.
#### Before data has 180 rows, now data has 103 rows.

### Merging the rows

#### when Postcode and Borough are same but Neighbourhoods are different, then neighbourds will be seperated by ','.
#### So Postcode and Borough pair can appear only once in the data, so we'll make a dictionary with (Postcode, Borough) as keys and the values will be corresponding neighbourhood/s.
#### Creating a dataframe using keys and values of dictionary

In [8]:
postcodes = df['PostalCode'].values
boroughs = df['Borough'].values
neighs = df['Neighbourhood'].values

#create a dictionary with keys as Postcode and Borough, keys of dictioaries are unique
dic = dict({(key1,key2): [] for key1, key2 in zip(postcodes, boroughs)})
print('Number of keys in the dictionary are: ', len(dic.keys()))

#filling the values of keys of dictionary
for postcode, borough, neigh in zip(postcodes,boroughs, neighs):
    key = (postcode, borough)
    dic[key].append(neigh)

df = pd.DataFrame(columns = ['Postal Code', 'Borough', 'Neighbourhood'])
for key, value in dic.items():
    postcode, borough, neig = key[0], key[1], value
    neig = ', '.join(neig)
    df = df.append({'Postal Code': postcode,
                     'Borough': borough,
                     'Neighbourhood': neig}, ignore_index = True)
print('Shape of final data is: ', df.shape)
df.head(10)

Number of keys in the dictionary are:  103
Shape of final data is:  (103, 3)


Postal Code           Borough                                 Neighbourhood
0         M3A        North York                                     Parkwoods
1         M4A        North York                              Victoria Village
2         M5A  Downtown Toronto                    Regent Park / Harbourfront
3         M6A        North York             Lawrence Manor / Lawrence Heights
4         M7A  Downtown Toronto  Queen's Park / Ontario Provincial Government
5         M9A         Etobicoke                              Islington Avenue
6         M1B       Scarborough                               Malvern / Rouge
7         M3B        North York                                     Don Mills
8         M4B         East York              Parkview Hill / Woodbine Gardens
9         M5B  Downtown Toronto                      Garden District, Ryerson

# Segmenting and Clustering PART 2

In [9]:
# import geocoder # import geocoder

# initialize your variable to None
# lat_lng_coords = None

# loop until you get the coordinates
# while(lat_lng_coords is None):
# g = geocoder.google('{}, Toronto, Ontario'.format(postal_code))
# lat_lng_coords = g.latlng

# latitude = lat_lng_coords[0]
# longitude = lat_lng_coords[1]


!wget http://cocl.us/Geospatial_data
try:
    df['Latitude'] = lats
    df['Longitude'] = lons
except:
    latlon = pd.read_csv('Geospatial_data')
    df = pd.merge(df, latlon, how= 'inner', on = 'Postal Code')
    # reorder column names and show the dataframe
    #df = df[['Postcode', 'Borough', 'Neighbourhood', 'Latitude', 'Longitude']]
print(df.shape)
df.head(10)

--2020-04-05 11:03:02--  http://cocl.us/Geospatial_data
Resolving cocl.us (cocl.us)... 158.85.108.86, 169.48.113.194, 158.85.108.83
Connecting to cocl.us (cocl.us)|158.85.108.86|:80... connected.
HTTP request sent, awaiting response... 301 Moved Permanently
Location: https://cocl.us/Geospatial_data [following]
--2020-04-05 11:03:02--  https://cocl.us/Geospatial_data
Connecting to cocl.us (cocl.us)|158.85.108.86|:443... connected.
HTTP request sent, awaiting response... 301 Moved Permanently
Location: https://ibm.box.com/shared/static/9afzr83pps4pwf2smjjcf1y5mvgb18rr.csv [following]
--2020-04-05 11:03:03--  https://ibm.box.com/shared/static/9afzr83pps4pwf2smjjcf1y5mvgb18rr.csv
Resolving ibm.box.com (ibm.box.com)... 107.152.27.197, 107.152.26.197
Connecting to ibm.box.com (ibm.box.com)|107.152.27.197|:443... connected.
HTTP request sent, awaiting response... 301 Moved Permanently
Location: /public/static/9afzr83pps4pwf2smjjcf1y5mvgb18rr.csv [following]
--2020-04-05 11:03:04--  https://ib

Postal Code           Borough                                 Neighbourhood  \
0         M3A        North York                                     Parkwoods   
1         M4A        North York                              Victoria Village   
2         M5A  Downtown Toronto                    Regent Park / Harbourfront   
3         M6A        North York             Lawrence Manor / Lawrence Heights   
4         M7A  Downtown Toronto  Queen's Park / Ontario Provincial Government   
5         M9A         Etobicoke                              Islington Avenue   
6         M1B       Scarborough                               Malvern / Rouge   
7         M3B        North York                                     Don Mills   
8         M4B         East York              Parkview Hill / Woodbine Gardens   
9         M5B  Downtown Toronto                      Garden District, Ryerson   

    Latitude  Longitude  
0  43.753259 -79.329656  
1  43.725882 -79.315572  
2  43.654260 -79.360636  
3  43.718518 -79.464763  
4  43.662301 -79.389494  
5  43.667856 -79.532242  
6  43.806686 -79.194353  
7  43.745906 -79.352188  
8  43.706397 -79.309937  
9  43.657162 -79.378937

# Segmenting and Clustering PART 3

#### Create a map of Toronto with neighbourhood superimposed on it

In [10]:
address = 'Toronto, Ontario'
geolocator = Nominatim(user_agent="ny_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of Toronto, Ontario are {}, {}.'.format(latitude, longitude))

The geograpical coordinate of Toronto, Ontario are 43.6534817, -79.3839347.


In [11]:
# create map of New York using latitude and longitude values
map_toronto = folium.Map(location=[latitude, longitude], zoom_start=10)

# add markers to map
for lat, lng, borough, neighborhood in zip(df['Latitude'], df['Longitude'], df['Borough'], df['Neighbourhood']):
    label = '{}, {}'.format(neighborhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_toronto)  
    
map_toronto

In [12]:
print('Total number of Borough = ', len(df['Borough'].unique()))

Total number of Borough =  10


#### Explore one borough, Downtown Toronto, then we'll do the same for other boroughs

In [13]:
downtown_toronto = df[df['Borough'] == 'Downtown Toronto'].reset_index(drop=True)
print(downtown_toronto.shape)
downtown_toronto.head()

(19, 5)


Postal Code           Borough                                 Neighbourhood  \
0         M5A  Downtown Toronto                    Regent Park / Harbourfront   
1         M7A  Downtown Toronto  Queen's Park / Ontario Provincial Government   
2         M5B  Downtown Toronto                      Garden District, Ryerson   
3         M5C  Downtown Toronto                                St. James Town   
4         M5E  Downtown Toronto                                   Berczy Park   

    Latitude  Longitude  
0  43.654260 -79.360636  
1  43.662301 -79.389494  
2  43.657162 -79.378937  
3  43.651494 -79.375418  
4  43.644771 -79.373306

#### Let's visualize the neighbourhood of Downtown Toronto in map

In [16]:
address = 'Downtown Toronto ,Toronto, Ontario'
geolocator = Nominatim(user_agent="ny_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude

# create map of Downtown Toronto using latitude and longitude values
map_downtown = folium.Map(location=[latitude, longitude], zoom_start= 11)

# add markers to map
for lat, lng, borough, neighborhood in zip(downtown_toronto['Latitude'], downtown_toronto['Longitude'], 
                                           downtown_toronto['Borough'], downtown_toronto['Neighbourhood']):
    label = '{}, {}'.format(neighborhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_downtown)  
    
map_downtown

# Explore the venues using Forsquare API

#### Let's explore Neighbourhood of Downtown Toronto in our Dataframe

In [17]:
#definig the latitude and longitude using above dataframe
lat = downtown_toronto.loc[0, 'Latitude'] # neighborhood latitude value
lon = downtown_toronto.loc[0, 'Longitude'] # neighborhood longitude value

neighborhood_name = downtown_toronto.loc[0, 'Neighbourhood'] # neighborhood name
print('Latitude and longitude values of {} are {}, {}.'.format(neighborhood_name, lat, lon))

CLIENT_ID = 'Q0DLPFSUOVIJZH4JDQ2JTRXT4ZUAXFSARHAQDGZ4C1VFOA2I' # your Foursquare ID
CLIENT_SECRET = 'I4MNQ1BZEZOQNT21T11GRIINKBYEBICZAYFXHP0XJ1BFNXKV' # your Foursquare Secret
VERSION = '20180605' # Foursquare API version

LIMIT = 100
radius =1000
url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
    CLIENT_ID, CLIENT_SECRET, VERSION, lat,lon, radius, LIMIT)

# gettig the venues data form Forsquare API in json format
results = requests.get(url).json()
results

Latitude and longitude values of Regent Park / Harbourfront are 43.6542599, -79.3606359.


{'meta': {'code': 200, 'requestId': '5e89bda9f7706a001b5a6925'},
 'response': {'suggestedFilters': {'header': 'Tap to show:',
   'filters': [{'name': 'Open now', 'key': 'openNow'}]},
  'headerLocation': 'Corktown',
  'headerFullLocation': 'Corktown, Toronto',
  'headerLocationGranularity': 'neighborhood',
  'totalResults': 118,
  'suggestedBounds': {'ne': {'lat': 43.66325990900001,
    'lng': -79.3482199002972},
   'sw': {'lat': 43.64525989099999, 'lng': -79.37305189970282}},
  'groups': [{'type': 'Recommended Places',
    'name': 'recommended',
    'items': [{'reasons': {'count': 0,
       'items': [{'summary': 'This spot is popular',
         'type': 'general',
         'reasonName': 'globalInteractionReason'}]},
      'venue': {'id': '54ea41ad498e9a11e9e13308',
       'name': 'Roselle Desserts',
       'location': {'address': '362 King St E',
        'crossStreet': 'Trinity St',
        'lat': 43.653446723052674,
        'lng': -79.3620167174383,
        'labeledLatLngs': [{'label':

### By looking at the json output, we can see that items key and we get useful inforamtion using results['respnse']['groups'][0]['items']

In [18]:
venues = results['response']['groups'][0]['items']
venues_df = json_normalize(venues) # flatten JSON
venues_df.head()

reasons.count                                      reasons.items  \
0              0  [{'summary': 'This spot is popular', 'type': '...   
1              0  [{'summary': 'This spot is popular', 'type': '...   
2              0  [{'summary': 'This spot is popular', 'type': '...   
3              0  [{'summary': 'This spot is popular', 'type': '...   
4              0  [{'summary': 'This spot is popular', 'type': '...   

                       referralId  \
0  e-0-54ea41ad498e9a11e9e13308-0   
1  e-0-53b8466a498e83df908c3f21-1   
2  e-0-574c229e498ebb5c6b257902-2   
3  e-0-5612b1cc498e3dd742af0dc8-3   
4  e-0-4ad4c05ef964a520bff620e3-4   

                                    venue.categories  \
0  [{'id': '4bf58dd8d48988d16a941735', 'name': 'B...   
1  [{'id': '4bf58dd8d48988d1e0931735', 'name': 'C...   
2  [{'id': '52e81612bcbc57f1066b7a37', 'name': 'D...   
3  [{'id': '4bf58dd8d48988d1c4941735', 'name': 'R...   
4  [{'id': '4deefb944765f83613cdba6e', 'name': 'H...   

                   venue.id                     venue.location.address  \
0  54ea41ad498e9a11e9e13308                              362 King St E   
1  53b8466a498e83df908c3f21                              368 King St E   
2  574c229e498ebb5c6b257902                              461 Cherry St   
3  5612b1cc498e3dd742af0dc8                              573 King St E   
4  4ad4c05ef964a520bff620e3  btwn Front, Cherry, Gardiner & Parliament   

  venue.location.cc venue.location.city venue.location.country  \
0                CA             Toronto                 Canada   
1                CA             Toronto                 Canada   
2                CA             Toronto                 Canada   
3                CA             Toronto                 Canada   
4                CA             Toronto                 Canada   

  venue.location.crossStreet  ...  \
0                 Trinity St  ...   
1              at Trinity St  ...   
2                        NaN  ...   
3          at St Lawrence St  ...   
4                        NaN  ...   

                       venue.location.labeledLatLngs venue.location.lat  \
0  [{'label': 'display', 'lat': 43.65344672305267...          43.653447   
1  [{'label': 'display', 'lat': 43.65355870959944...          43.653559   
2  [{'label': 'display', 'lat': 43.65324910177244...          43.653249   
3  [{'label': 'display', 'lat': 43.65636850543279...          43.656369   
4  [{'label': 'display', 'lat': 43.65024435658077...          43.650244   

  venue.location.lng  venue.location.neighborhood  venue.location.postalCode  \
0         -79.362017                          NaN                    M5A 1K9   
1         -79.361809                          NaN                        NaN   
2         -79.358008                          NaN                    M5A 0H7   
3         -79.356980                          NaN                    M5A 4L3   
4         -79.359323                          NaN                    M5A 3C4   

  venue.location.state                        venue.name venue.photos.count  \
0                   ON                  Roselle Desserts                  0   
1                   ON                     Tandem Coffee                  0   
2                   ON            Cooper Koo Family YMCA                  0   
3                   ON                    Impact Kitchen                  0   
4                   ON  The Distillery Historic District                  0   

  venue.photos.groups  venue.venuePage.id  
0                  []                 NaN  
1                  []                 NaN  
2                  []                 NaN  
3                  []                 NaN  
4                  []                 NaN  

[5 rows x 22 columns]

#### We can see there are many details about the venues near first neighbourhood of Downtown Toronto.
#### But we're interested in only the venue name, categories, location (latitde and longitude).
#### Let's take only those columns which are needed.

In [19]:
# filter columns
cols = ['venue.name', 'venue.categories', 'venue.location.lat', 'venue.location.lng']
venues_df = venues_df.loc[:, cols]
# to show whole text in columns in dataframe python: https://stackoverflow.com/a/25352191
pd.set_option("display.max_colwidth" , -1)
venues_df.head()

venue.name  \
0  Roselle Desserts                   
1  Tandem Coffee                      
2  Cooper Koo Family YMCA             
3  Impact Kitchen                     
4  The Distillery Historic District   

                                                                                                                                                                                                                                                venue.categories  \
0  [{'id': '4bf58dd8d48988d16a941735', 'name': 'Bakery', 'pluralName': 'Bakeries', 'shortName': 'Bakery', 'icon': {'prefix': 'https://ss3.4sqi.net/img/categories_v2/food/bakery_', 'suffix': '.png'}, 'primary': True}]                                           
1  [{'id': '4bf58dd8d48988d1e0931735', 'name': 'Coffee Shop', 'pluralName': 'Coffee Shops', 'shortName': 'Coffee Shop', 'icon': {'prefix': 'https://ss3.4sqi.net/img/categories_v2/food/coffeeshop_', 'suffix': '.png'}, 'primary': True}]                         
2  [{'id': '52e81612bcbc57f1066b7a37', 'name': 'Distribution Center', 'pluralName': 'Distribution Centers', 'shortName': 'Distributor', 'icon': {'prefix': 'https://ss3.4sqi.net/img/categories_v2/building/default_', 'suffix': '.png'}, 'primary': True}]        
3  [{'id': '4bf58dd8d48988d1c4941735', 'name': 'Restaurant', 'pluralName': 'Restaurants', 'shortName': 'Restaurant', 'icon': {'prefix': 'https://ss3.4sqi.net/img/categories_v2/food/default_', 'suffix': '.png'}, 'primary': True}]                               
4  [{'id': '4deefb944765f83613cdba6e', 'name': 'Historic Site', 'pluralName': 'Historic Sites', 'shortName': 'Historic Site', 'icon': {'prefix': 'https://ss3.4sqi.net/img/categories_v2/arts_entertainment/historicsite_', 'suffix': '.png'}, 'primary': True}]   

   venue.location.lat  venue.location.lng  
0  43.653447          -79.362017           
1  43.653559          -79.361809           
2  43.653249          -79.358008           
3  43.656369          -79.356980           
4  43.650244          -79.359323

In [20]:
# reset set option dataframe: https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.reset_option.html
pd.reset_option('display.max_colwidth')

### Name, location (lat and lng) looks good but we still need to etract the catgories.
### Categories is in venue.categories[0]['name']

In [21]:
venues_df['venue.categories'] = venues_df.apply(lambda x: x['venue.categories'][0]['name'], axis=1)
venues_df.head()

venue.name     venue.categories  venue.location.lat  \
0                  Roselle Desserts               Bakery           43.653447   
1                     Tandem Coffee          Coffee Shop           43.653559   
2            Cooper Koo Family YMCA  Distribution Center           43.653249   
3                    Impact Kitchen           Restaurant           43.656369   
4  The Distillery Historic District        Historic Site           43.650244   

   venue.location.lng  
0          -79.362017  
1          -79.361809  
2          -79.358008  
3          -79.356980  
4          -79.359323

#### Now every look good except the names of column let's change the name of columns

In [22]:
# clean columns
venues_df.columns = [col.split(".")[-1] for col in venues_df.columns]
print('{} Venues are returned for: {}'.format(venues_df.shape[0], neighborhood_name))
venues_df.head()

100 Venues are returned for: Regent Park / Harbourfront


name           categories        lat        lng
0                  Roselle Desserts               Bakery  43.653447 -79.362017
1                     Tandem Coffee          Coffee Shop  43.653559 -79.361809
2            Cooper Koo Family YMCA  Distribution Center  43.653249 -79.358008
3                    Impact Kitchen           Restaurant  43.656369 -79.356980
4  The Distillery Historic District        Historic Site  43.650244 -79.359323